### ISSUE TITLE
In Process hosting of PSMoveService

### ISSUE THREAD

**Issue by HipsterSloth** [cboulay/PSMoveService#171]
This is a low priority work item. Just an idea I wanted to log and open up for brainstorming. Some of this is a bit hand wave-y.

In another issue I had discussed trying to improve the latency of the PSMove SteamVR plugin by supporting a named pipe connection to the service when the service and the client are on the same machine.

It occurred to me that the better solution might just be for the SteamVR plugin (or any future psmove plugin) to support spawning it's own instance of the PSMoveService inside the plugin and then register that instance with the client interface. In the client API There would be two PSM_Initialize methods. The inproc version could take an interface pointer to the service. 

We already have the top level PSMoveService wrapped up in a class, it would just need to be split into two pieces. And I have another task to rip out our usage of Boost::Application. This work could be part of that refactor. The PSMoveService exe would just become a thin shell that manages service/daemon registration and uses a new PSMoveService DLL containing the new PSMoveServiceInstance class and interface. I think the DLL would just expose a C style interface for allocating, updating, and message handling.

The Client and Server Network manager would need to be put behind a message interface. Then InProc message interface would need to be implemented.

The nice thing about this is that you would no longer need to start up PSMoveService when launching SteamVR. It would just exist as part of vrserver.exe. And the InProc server will still have the network listener active so the config tool can still connect to it that way.

**Comment by cboulay** [comment]
In this scenario, would the InProc PSMoveService be in its own thread? If not, then this isn't very good for Unity or UE4 plugins because then the PSMoveService processing would be in the game thread. It's good for the SteamVR plugin because that's running an out-of-game process anyway.

However, before going down this road, I think it'll be worthwhile to spend some time minimizing interproc communication latency. If both service and client are running on the same machine then there are some other options. There's shared memory, like you said, but also the architecture and endianness is guaranteed to be the same so it's possible to not serialize the data and instead just pass binary blobs around in UDP ports. (I wonder how much of the latency is due to serialization itself and the fact that the serialized data is much larger than the binary data).

**Comment by HipsterSloth** [comment]
Yeah it would definitely have to be in it's own thread. Inside vrserver.exe each plugin already gets it's own thread to run in, so we would probably need to pass a flag to the InProc service construction about whether to run in a thread or not.

Agreed on the need for profiling the serialization. I suspect more time is spent just making all of the protobuf wrapper calls and not so much the endian swapping work.

The UDP packets are also always the same size regardless of the serialized data (500bytes + header size). This is because the client asio udp read api won't call back until the requested number of bytes are read from the socket. If we wanted variable UDP packet sizes we would need a more complicated message assembly mechanism. With lossy/out of order packets I'm not sure that approach would be a latency win over fixed size large packets or not. Something to look into though.

I think even when we get that serialization optimized there are still some machines that I think are buffering the UDP stream, though I need to implement some more logging to prove that this is the case and not something else.

**Comment by cboulay** [comment]
Make sure to look at [0MQ](http://zeromq.org/) and [nanomsg](http://nanomsg.org/). Seems like they have ways to easily transition between inproc and IPC (-non-network?) and network stacks.

**Comment by HipsterSloth** [comment]
Gave a brief look at nanomsg last night. I like the lightweight design of it and it looks like it supports non-blocking io. And it's MIT license. We'd still need to use protocol buffers with it since it just sends raw buffers. Still need to look over 0MQ, but I think the fact that it's LGPL might make it a non-starter since it would have to be in both the client and the server.

**Comment by cboulay** [comment]
LGPL is a lot better than GPL. LGPL allows proprietary code using the software to remain closed-source as long as it is not a 'derivative work', and this is usually made obvious by linking against a dll instead of including the source directly. But this does add another DLL requirement to the client so that can be annoying.

I was hoping that the easy-switching between inproc and other transport mechanisms would also facilitate switching between no-serialization and protobuf serialization. I haven't looked at the library enough to know if it would help there. If not, then there might not be any performance gains, but it's still probably easier to use than boost.asio

**Comment by HipsterSloth** [comment]
LGPL is definitely a lot better than GPL, but unfortunately many companies will simply refuse to associate with code that uses the LGPL because of legal paranoia. My last company I worked at had that policy. It even makes investors nervous. As part of our current investment deal we had to sign a disclosure document outlining any GPL or LGPL code we had in our code base (none currently). It doesn't mean that they won't invest, but it's a red flag apparently. And to be clear, I agree that it should be fine, but the unfortunate practical reality is that if we link LGPL'ed code into the client API there will be a subset of people that can't use this service anymore. 

I'm totally down for the the easy switching between named pipe/socket. Also getting rid of the need for boost.asio would definitely help build times. As for faster serialization we might also look at [CapNProto](https://capnproto.org/install.html), made by one of the authors of google protocol buffers.
